### DISCUSSION TITLE
using tanstack/reat-query or providing an alternative to use it

### DISCUSSION THREAD

**Discussion by sharmapukar217** [rakkasjs/rakkasjs#65] [Ideas]
I checked out your example of using react query but that couldn't be used with useSSQ or useSSM ( correct me if I'm wrong ). Why I like react query is it does have all or atleast necessary tools like devtools and cache persister. 
Alternatively an option for cache persister can be provided while devtool  could be ported i guess. I saw the code and i think it won't be much difficult to port

**Comment by cyco130** [comment]
Soon I will make the compiler customizable so you can create your own magical `useServerSideXxx` functions. In the meantime you can use `runServerSideQuery` and `runServerSideMutation` (not yet documented):

```ts
const { requestContext } = usePageContext();
const fetched = useYourFavoriteDataFetchingHook(async () => {
  const result = await runServerSideQuery(requestContext, () => {
    // Do some server-side work here
    // ...
  });

  // Do something with the result here
  // ...

  return result; // or whatever
});
```

`runServerSideMutation` is similar and doesn't require you to pass the `requestContext` argument but it can only be used on the client, not during SSR.

BTW there's a bug in the react-query demo, I'll fix it soon.

**Comment by sharmapukar217** [comment]
meanwhile can i try to make the devtools work with rakkasjs (if possible) ?

**Comment by cyco130** [comment]
I'm still undecided about whether to abandon Rakkas's own `useQuery` for now because it's Suspense-only which makes it a lot smaller.

But it would be an argument in favor of keeping it if react-query dev tools worked with it. I'd probably extract into its own package actually.

**Reply by sharmapukar217** [reply]
> useQuery

i guess keeping query and mutation function into separate package will be better because there might be some who just need api routes and page routes but not these functions. Or maybe some might want to implement graphql, and keeping useQuery within rakkasjs will be no use.

**Reply by sharmapukar217** [reply]
sth like @rakkas/rpc

**Reply by cyco130** [reply]
It may not be like that exactly. Rakkas has no other way of data fetching, we need _some_thing out of the box for it to be of any use.

**Reply by sharmapukar217** [reply]
Agreed with this

**Reply by sharmapukar217** [reply]
Providing data fetching out of the box is a plus point
